lib/spack/spack/test/conftest.py

```diff
@@ -1234,3 +1239,18 @@ def mock_test_stage(mutable_config, tmpdir):
     mutable_config.set('config:test_stage', tmp_stage)
 
     yield tmp_stage
+
+
+@pytest.fixture(scope='function')
+def patch_git_ref(monkeypatch):
+
+    def from_ref(ref):
+        assert isinstance(ref, fs.GitRef), ref
+
+        def from_directive(*args, **kwargs):
+            return ref
+        # py2 complains unless you bind the .from_directive() classmethod to the class.
+        from_directive = from_directive.__get__(fs.GitRef, fs.GitRef.__class__)
+        monkeypatch.setattr(fs.GitRef, 'from_directive', from_directive)
+
+    return from_ref
```

Comment:
Yes, this is extremely difficult to follow. This would not be necessary at all if we ensured fetchers used in testing were always constructed with arguments sufficient to make `.from_directive()` pass. Now that I've gotten a handle on the unit tests passing, I think we can probably remove this fixture entirely and instead just construct fetch strategies with arguments that are sufficient to disambiguate the git ref they refer to. I want to continue raising a `FetcherConflict` in the `GitFetchStrategy` constructor, because erroring out earlier is always nicer.

I will try to just remove this fixture entirely. I think the result will be much easier to follow.

Reply:
Ok, I reconsidered this, and the reason I would prefer not to do:
> Maybe something like `mock_git_ref_factory`, with a parameter "type" to choose among 'branch', 'tag', or 'commit'.?

is because to me that gets annoyingly ambiguous pretty fast, and the type `GitRef` was intended to be part of the solution to this general problem, by ensuring that none of our code (including tests) are relying on ambiguous behavior.

For example, I learned that if I set `init.defaultBranch` to `main` globally, all of our git tests fail, because they read from the user's global git config (*not* the one from the spack repo itself)! So I would really prefer to funnel test cases through the `GitRef` constructor to avoid this kind of issue.

However, I have renamed it to `patch_from_directive_for_git_ref` as you suggested above.